pkg/webhook/certs.go

```diff
@@ -19,24 +19,74 @@ package webhook
 import (
 	"crypto/tls"
 	"io/ioutil"
+	"time"
+
+	"github.com/golang/glog"
 )
 
-// certBundle is a container of a X509 certificate file and a corresponding key file for the
+// certProvider is a container of a X509 certificate file and a corresponding key file for the
 // webhook server, and a CA certificate file for the API server to verify the server certificate.
-type certBundle struct {
+type certProvider struct {
 	serverCertFile string
 	serverKeyFile  string
 	caCertFile     string
+	reloadInterval time.Duration
+	ticker         *time.Ticker
+	stopChannel    chan interface{}
+	currentCert    *tls.Certificate
 }
 
-// configServerTLS configures TLS for the admission webhook server.
-func configServerTLS(certBundle *certBundle) (*tls.Config, error) {
-	cert, err := tls.LoadX509KeyPair(certBundle.serverCertFile, certBundle.serverKeyFile)
+func NewCertProvider(serverCertFile, serverKeyFile, caCertFile string, reloadInterval time.Duration) (*certProvider, error) {
+	cert, err := tls.LoadX509KeyPair(serverCertFile, serverKeyFile)
 	if err != nil {
 		return nil, err
 	}
+	return &certProvider{
+		serverCertFile: serverCertFile,
+		serverKeyFile:  serverKeyFile,
+		caCertFile:     caCertFile,
+		reloadInterval: reloadInterval,
+		currentCert:    &cert,
+	}, nil
+}
+
+func (c *certProvider) Start() {
+	stopChannel := make(chan interface{})
+	c.stopChannel = stopChannel
+	ticker := time.NewTicker(c.reloadInterval)
+	c.ticker = ticker
+	go func() {
+		for {
+			select {
+			case <-stopChannel:
+				return
+			case <-c.ticker.C:
+				c.updateCert()
+			}
+		}
+	}()
+}
 
-	return &tls.Config{Certificates: []tls.Certificate{cert}}, nil
+func (c *certProvider) tlsConfig() *tls.Config {
+	return &tls.Config{
+		GetCertificate: func(ch *tls.ClientHelloInfo) (*tls.Certificate, error) {
+			return c.currentCert, nil
```

Comment:
Do we need to use a mutex to guard `currentCert`, which may be read in `tlsConfig` and written in `updateCert`?

Reply:
It's probably a good idea, since this behavior is technically undefined (although AFAICT it's safe on all modern architectures).